draft-ietf-quic-transport.md

```diff
@@ -4099,11 +4102,13 @@ preferred_address (0x000d):
   the end of the handshake, as described in {{preferred-address}}.  The format
   of this transport parameter is the PreferredAddress struct shown in
   {{fig-preffered-address}}.  This transport parameter is only sent by a server.
+  Servers MAY choose to only send a preferred address of one address family by
+  sending a zero-length address for the other family.
 
 ~~~
    struct {
-     enum { IPv4(4), IPv6(6), (15) } ipVersion;
-     opaque ipAddress<4..2^8-1>;
+     opaque ipv4Address<0..2^8-1>;
+     opaque ipv6Address<0..2^8-1>;
      uint16 port;
```

Comment:
I'm fine with only supporting v4 and v6.

Reply:
And I'm ok with having the same port across versions.  Which all suggests a smaller change:
```
ipv4Address[4];
ipv6Address[16];
uint16 port;
```

Reply:
Why would the ipv4 and ipv6 port numbers be the same? That seems like a special case derived from the HTTP use case.

Reply:
If there's a real use case for separate port numbers, I can see adding that flexibility, but otherwise I prefer @martinthomson 's suggestion.

Reply:
p2p applications might not assign any meaning to ports, but just use a random port number. There's no reason why those values would be the same for IPv4 and IPv6.

Reply:
Yes, this seems like an unnecessary constraint. Especially since the mechanism can be used to advertise non-standard ports. Random ports could also be a first  line of defence against DoS attacks and they could be used to move endpoints dynamically over time. Forcing two different interfaces to be in sync isn't ideal for that.

Reply:
This might be a dumb question, but would P2P applications use the feature, considering that the intent is to remap clients connecting to an anycast address to a unicast address?

Reply:
I think the intent is to deal with anycast, but I could see it being used defensively against attacks.

As for P2P there is a strong case for both ends to use a known fixed port. However, I could also see this as a way to load balance multiple connections to multiple threads or processes.

Reply:
@marten-seemann most implementations (on the operating systems I'm familiar with) will use a single dual-stack listening socket for all address families so unless you go out of your way you will end up with the same port. That's true of HTTP servers and P2P apps.

That said, if there is consensus that separate ports are useful, I'm OK adding separate ports to this PR.

Reply:
I'd err on the side of making it as generic as possible: that is, separate ports.

Reply:
The latest version has separate ports, see the diff.